Comorphisms/SuleCFOL2TPTP.hs

```diff
@@ -0,0 +1,682 @@
+import Data.Char
+import Data.Function
+import Numeric (showHex)
+
+import CASL.Logic_CASL
+import CASL.AS_Basic_CASL as CAS
+import CASL.Cycle
+import CASL.Sublogic as SL
+import CASL.Sign as CSign
+import CASL.MapSentence
+import CASL.Morphism
+import CASL.Quantification
+import CASL.Overload
+import CASL.Utils
+import CASL.Inject
+import CASL.Simplify
+import CASL.ToDoc
+
+import TPTP.AS as TAS
+import TPTP.Morphism as TMorphism
+import TPTP.Sign as TSign
+import TPTP.Logic_TPTP
+import TPTP.Sublogic
+
+data TPTP_FOF = TPTP_FOF
+
+instance Show TPTP_FOF where
+  show TPTP_FOF = "TPTP_FOF"
+
+-- | The identity of the comorphisms
+data GenSuleCFOL2TPTP a = GenSuleCFOL2TPTP a deriving Show
+
+suleCFOL2TPTP :: GenSuleCFOL2TPTP TPTP_FOF
+suleCFOL2TPTP = GenSuleCFOL2TPTP TPTP_FOF
+
+-- | TPTP theories
+type TPTPTheory = (TSign.Sign, [Named Sentence])
+
+instance Show a => Language (GenSuleCFOL2TPTP a) where
+  language_name (GenSuleCFOL2TPTP a) = "CASL2" ++ show a
+
+instance Show a => Comorphism (GenSuleCFOL2TPTP a)
+               CASL CASL_Sublogics
+               CASLBasicSpec CASLFORMULA SYMB_ITEMS SYMB_MAP_ITEMS
+               CASLSign
+               CASLMor
+               CSign.Symbol RawSymbol ProofTree
+               TPTP.Logic_TPTP.TPTP Sublogic TAS.BASIC_SPEC Sentence () ()
+               TSign.Sign TMorphism.Morphism TSign.Symbol () ProofTree where
+    sourceLogic (GenSuleCFOL2TPTP _) = CASL
+    sourceSublogic (GenSuleCFOL2TPTP a) = SL.cFol
+                      { sub_features = LocFilSub
+                      , cons_features = NoSortGen
+                      , has_empty_sorts = show a == show TPTP_FOF }
+    targetLogic (GenSuleCFOL2TPTP _) = TPTP.Logic_TPTP.TPTP
+    mapSublogic cid sl = Just FOF
+    map_theory (GenSuleCFOL2TPTP a) = transTheory
+    has_model_expansion (GenSuleCFOL2TPTP _) = True
+
+transTheory :: (FormExtension f, Eq f)
+            => (CSign.Sign f e, [Named (FORMULA f)])
+            -> Result TPTPTheory
+transTheory (sign, sens) = do
+  let signWithRenamings@(preparedSign, _, _) = prepareSign sign
+  let preparedSens = map prepareNamedFormula sens
+  (tptpSign, signSentences) <- translateSign preparedSign
+  translatedSentences <- mapM (translateNamedFormula signWithRenamings) preparedSens
+  return (tptpSign, signSentences ++ translatedSentences)
+
+prepareNamedFormula :: (FormExtension f, Eq f)
+                    => Named (FORMULA f) -> Named (FORMULA f)
+prepareNamedFormula formula =
+  let expandedFormula =
+        codeOutConditionalF id $ -- Expand conditionals
+        codeOutUniqueExtF id id $ -- Expand "exists!" quantification
+        sentence formula
+  in  formula { sentence = expandedFormula }
+
+translateNamedFormula :: (FormExtension f, Eq f)
+                      => SignWithRenamings f e
+                      -> Named (FORMULA f) -> Result (Named TSign.Sentence)
+translateNamedFormula signWithRenamings x = do
+  let nameS = "sentence_" ++ toAlphaNum (map toLower $ senAttr x)
```

Comment:
When I implemented this, it was needed because of Ontohub's Loc/Id restriction: Sentences and symbols share the same namespace. With the new Ontohub implementation, this not necessary.

If this won't be applied to the old Ontohub, I can remove the prefix.

Reply:
yes, then please remove it.